src/libraries/System.Net.Http/src/System/Net/Http/HttpMessageHandler.cs

```diff
@@ -17,6 +17,15 @@ protected HttpMessageHandler()
             if (NetEventSource.IsEnabled) NetEventSource.Info(this);
         }
 
+        // We cannot add abstract member to a public class in order to not to break already established contract of this class.
+        // So we add virtual method, override it everywhere internally and provide proper implementation.
+        // Unfortunately we cannot force everyone to implement so in such case we have no other option that to do sync-over-async.
+        protected internal virtual HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Info(this, $"Doing sync-over-async due to lack of {nameof(Send)} override");
+            return SendAsync(request, cancellationToken).GetAwaiter().GetResult();
```

Comment:
I know we've talked about this blocking approach in the base methods, and this is what I had in my prototype, but I'm having second thoughts.

The scenarios for which we're adding these synchronous methods arelimited.  We want to make it possible to use the built-in HTTP stack for synchronous operations, but we're not going out of our way to promote it nor exposing lots of easy helpers to get to it.  We're doing this primarily to enable developers that are forced to use synchrony to get the best possible scalability and performance in doing so, but not only is the resulting scale for sync worse than for async, falling back to sync-over-async will not only be worse than that, it may possibly be worse than what the developer could achieve if they were managing it themselves knowing that it was going to be sync-over-async.

This makes me think we should be more constrained in what we "support".  We'll be forced to fall back to sync-over-async if:
- Any custom HttpContent or HttpMessageHandler is in the mix that hasn't overridden the relevant methods.
- A protocol other than HTTP/1.x is used.

So... what if we just simply threw in those situations?  We wouldn't need to worry about race conditions in general here, as the construction of the client/handler and messages sent would dictate whether it succeeded or failed, and if it fails, it'll do so with a clear message about why and what the developer would need to do to succeed, e.g. if you set HTTP/2 on your message, either use the recommended async methods or change your protocol to be 1.1 if you must use sync.

This would mean that we:
- Change HttpMessageHandler.Send to unconditionally throw a NotSupportedException.
- Change HttpContent.SerializeToStream to unconditionally throw a NotSupportedException.
- Change SocketsHttpHandler.Send to throw a NotSupportedException _if_ the HttpRequestMessage.Version > 1.1

Everything else would remain the same; in particular, devs won't see the NotSupportedExceptions from the base HttpMessageHandler/HttpContents if only using the built-in handlers and content types, because they all override these.

The more I think about the ecosystem impact to implicitly trying to support these effectively unsupported situations by blocking, the more I think we'll be digging ourselves a deep hole.  Better to only support the core pieces we can reason about and make promises about from a scale perspective, and then in the future if necessary consider loosening those constraints.

@scalablecory, @davidsh, @KrzysztofCwalina, @davidfowl, thoughts?

Reply:
Your proposal seems reasonable. One thing that could be a sticking point is that the person who creates the derived class isn't necessarily the same person who will see the exception. So if we assume the caller doesn't have the ability to modify the implementation, would the message be enough to unblock the scenarios the caller might care about?

On the implementers' side, would an analyzer (or similar) be appropriate? "Hey, you're creating a component that will fail in _xyz_ scenarios."

Reply:
In short, making the sync code path opt in rather than defaulting to sync-over-async. Sounds good to me.

Reply:
> In short, making the sync code path opt in rather than defaulting to sync-over-async. Sounds good to me.

Yup.

> would the message be enough to unblock the scenarios the caller might care about?

Luckily we get to write the message :smile:

> On the implementers' side, would an analyzer (or similar) be appropriate?

It could be.  We have this issue in a variety of places, where we add new virtual methods to a type and we really want types to override them (we'd make them abstract if we could).

That said, part of my suggestion here is that we recognize that async really is the desired path, and we don't want to be forcing everyone to write a lot of code to deal with the undesirable one.  So, we make everything in the box that can work well in fact work well, and then it ends up being optional whether you support sync.  If a consumer of an HttpClient instances wants to use the sync methods, but the HttpClient was constructed with components that don't support it, rather than just getting silently bad behaviors, it'll be in their face, and the consumer and the creator of the HttpClient (if they're not the same people) can sit down and chat about how best to address the gap.

Reply:
This makes me very happy!

Reply:
> Better to only support the core pieces we can reason about and make promises about from a scale perspective, and then in the future if necessary consider loosening those constraints.

If you feel like there are too many unknowns to do this reasonably well then I am fine with this, but it sucks from a usability perspective.

Are we optimizing for the right user here?

We initially talked about adding this API for the rare user who wants perf/scalability but also wants sync, but I suspect (strongly hope) that this is actually a small group. I expect the biggest user of this to be the one who doesn't care about or doesn't understand scalability, and wants something that "just works". If the API has a bunch of non-corner cases where it throws, then this is far from "just works" and fails those users.

Perhaps we should rethink the API, making it harder to find (maybe require reflection?) so that you'd only ever use it if you are in the 1st scenario, not the 2nd.

Reply:
There are two use cases here:
1. Wanting to make it super-simple to do everything synchronously.
2. Wanting to make it possible for more advanced consumption to avoid sync-over-async when necessary.

We've already chosen (2); if we were doing (1), we would be adding a ton more surface area to HttpClient itself, rather than a single Send method.  And for (2), the pit of success is knowing that things are actually synchronous, so throwing NotSupportedException in the major cases we can't make that happen (custom types that haven't overridden what they need to, HTTP/2, IConnectionFactory, etc.) will IMO better guide developers to the right results.

We can always remove the constraints later if desired.  We can't reasonably add them later.

Reply:
Resolved by throwing NSE as agreed in discussion bellow.